File: reference/python/nexalert_reference/reliability.py
"""
Reliability (R_i) reference implementation.

Specification: Document 04, Section 3.3
Formula: R_i = H_i × Q_i × K_i
Range: [0, 1]
"""
import numpy as np
from typing import Optional


def compute_reliability(
    h_i: Optional[float],
    q_i: Optional[float],
    k_i: Optional[float]
) -> Optional[float]:
    """
    Compute evidence reliability R_i.

    Args:
        h_i: Sensor health H_i [0, 1] or None
        q_i: Signal quality Q_i [0, 1] or None
        k_i: Calibration validity K_i [0, 1] or None

    Returns:
        R_i in [0, 1], or None if any input is None

    Specification: Document 04, Section 3.3

    Invariants:
        - R_i = 0 if any component is 0 (strict trust gate, Doc 04 Sec 3.3)
        - R_i is NOT double-counted in confidence later (Doc 04 Sec 3.3)
        - K_i is explicit input; NO hard-coded default (calibration from Phase 6)
        - None inputs preserve missing != zero (IMPLEMENTATION_CONSTITUTION.md Sec 3)
    """
    # Preserve missing != zero
    if h_i is None or q_i is None or k_i is None:
        return None

    # Validate inputs (Doc 04 Sec 3.3: all components in [0,1])
    if not (0.0 <= h_i <= 1.0):
        raise ValueError(f"h_i={h_i} out of range [0,1] per Doc 04 Sec 3.3")
    if not (0.0 <= q_i <= 1.0):
        raise ValueError(f"q_i={q_i} out of range [0,1] per Doc 04 Sec 3.3")
    if not (0.0 <= k_i <= 1.0):
        raise ValueError(f"k_i={k_i} out of range [0,1] per Doc 04 Sec 3.3")

    # Multiplicative trust gate (Doc 04 Sec 3.3)
    r_i = h_i * q_i * k_i

    # Defensive clamp
    r_i = np.clip(r_i, 0.0, 1.0)

    return float(r_i)

File: firmware/components/intelligence/reliability.c
/**
 * @file reliability.c
 * @brief Evidence reliability (R_i) calculation for NexAlert edge intelligence
 *
 * Specification: Document 04, Section 3.3
 * Reference: reference/python/nexalert_reference/reliability.py
 *
 * Formula: R_i = H_i × Q_i × K_i
 */

#include "reliability.h"
#include <math.h>

reliability_result_t compute_reliability(
    float h_i,
    float q_i,
    float k_i
)
{
    reliability_result_t result = {
        .r_i = 0.0f,
        .valid = false
    };

    // Preserve missing != zero (Python: if h_i is None or q_i is None or k_i is None)
    if (isnan(h_i) || isnan(q_i) || isnan(k_i)) {
        result.valid = false;
        return result;
    }

    // Validate h_i range [0, 1] (Doc 04 Sec 3.3)
    if (h_i < 0.0f || h_i > 1.0f) {
        result.valid = false;
        return result;
    }

    // Validate q_i range [0, 1] (Doc 04 Sec 3.3)
    if (q_i < 0.0f || q_i > 1.0f) {
        result.valid = false;
        return result;
    }

    // Validate k_i range [0, 1] (Doc 04 Sec 3.3)
    if (k_i < 0.0f || k_i > 1.0f) {
        result.valid = false;
        return result;
    }

    // Multiplicative trust gate (Doc 04 Sec 3.3)
    // R_i = 0 if any component is 0 (strict trust gate)
    float r_i = h_i * q_i * k_i;

    // Defensive clamp to [0, 1]
    if (r_i < 0.0f) {
        r_i = 0.0f;
    } else if (r_i > 1.0f) {
        r_i = 1.0f;
    }

    result.r_i = r_i;
    result.valid = true;
    return result;
}
